spec.md

```diff
@@ -778,6 +800,182 @@ message Volume {
+  // permitted =
+  //   {"region": "R1", "zone": "Z2"},
+  //   {"region": "R1", "zone": "Z3"}
+  // preferred =
+  //   {"region": "R1", "zone": "Z3"}
+  // then the the SP SHOULD first attempt to make the provisioned volume
+  // available from "zone" "Z3" in the "region" "R1" and fall back to
+  // "zone" "Z2" in the "region" "R1" if that is not possible.
+  //
+  // Example 2:
+  // Given a volume should be accessible from a single zone, and
+  // permitted =
+  //   {"region": "R1", "zone": "Z2"},
+  //   {"region": "R1", "zone": "Z3"},
+  //   {"region": "R1", "zone": "Z4"},
+  //   {"region": "R1", "zone": "Z5"}
+  // preferred =
+  //   {"region": "R1", "zone": "Z4"},
+  //   {"region": "R1", "zone": "Z2"}
+  // then the the SP SHOULD first attempt to make the provisioned volume
+  // accessible from "zone" "Z4" in the "region" "R1" and fall back to
+  // "zone" "Z2" in the "region" "R1" if that is not possible. If that
+  // is not possible, the SP may choose between either the "zone"
+  // "Z3" or "Z5" in the "region" "R1".
+  //
+  // Example 3:
+  // Given a volume should be accessible from TWO zones, and
+  // permitted =
+  //   {"region": "R1", "zone": "Z2"},
+  //   {"region": "R1", "zone": "Z3"},
+  //   {"region": "R1", "zone": "Z4"},
+  //   {"region": "R1", "zone": "Z5"}
+  // preferred =
+  //   {"region": "R1", "zone": "Z5"},
+  //   {"region": "R1", "zone": "Z3"}
+  // then the the SP SHOULD first attempt to make the provisioned volume
+  // accessible from the combination of the two "zones" "Z5" and "Z3" in
+  // the "region" "R1". If that's not possible, it should fall back to
+  // a combination of "Z5" and other possibilities from the list of
+  // permitted. If that's not possible, it should fall back  to a
+  // combination of "Z3" and other possibilities from the list of
+  // permitted. If that's not possible, it should fall back  to a
+  // combination of other possibilities from the list of permitted.
+  repeated Topology preferred = 2;
+}
+
+// Topology is a map of topological domains to topological segments.
+// A topological domain is a sub-division of a cluster, like "region",
+// "zone", "rack", etc.
+// A topological segment is a specific instance of a topological domain,
+// like "zone3", "rack3", etc.
+// For example {"zone": "Z1", "rack": "R3"}
+// Each key (topological domain) must be 63 characters or less and
+// consist of alphanumeric characters, '-', '_' or '.'.
+// Each key SHOULD include the plugin's host company name and the plugin
+// name, to minimize the possibility of collisions with keys from other
+// plugins.
+// Keys MUST be case-insensitive. Meaning the keys "Zone" and "zone"
+// MUST not both exist.
+// Each value (topological segment) MUST contain 1 or more strings.
+// Each string MUST be 63 characters or less and begin and end with an
+// alphanumeric character with '-', '_', '.', or alphanumerics in
+// between.
+message Topology {
```

Comment:
I was wondering if this could be presented in `ListVolumesResponse` for CO to make scheduling decisions.

Also, should the snapshots be aware of topology as well?

Reply:
> I was wondering if this could be presented in ListVolumesResponse for CO to make scheduling decisions.

It will be since it is part of `message Volume`.

> Also, should the snapshots be aware of topology as well?

I thought about it, but I don't think any of the snapshoting functionality (CreateSnapshot, DeleteSnapshot) needs it.

Reply:
Oh year you're right. I apologize for not reading it carefully ;)

Reply:
Re: Topology awareness for snapshots. This means that if CO want to create a volume without expecting an error due to topology, it needs to discover the topology of the source volume (though `ListVolumes` or `ValidateVolumeCapabilities`), assuming the snapshot has the same topology requirements as its source volume? Is this a reasonable assumption?

Reply:
It's probably worth stating that assumption explicitly in the spec if that's the direction we're headed vs. leaving it vague and unspecified.

Reply:
Done: https://github.com/container-storage-interface/spec/issues/241